### DISCUSSION TITLE
How can I bypass reCaptcha v2 in cypress?

### DISCUSSION THREAD

**Discussion by sejal-bansal** [cypress-io/cypress#29923] [Questions and Help]
I am very new to cypress and have been trying to solve this problem, and nothing seems to work. Can someone please help me? 
/// <reference types = "cypress" />

it('Login test', function() {
    cy.visit('https://', {
        onBeforeLoad(win) {
            win.grecaptcha = {
                ready: (callback) => {
                    callback();
                },
                execute: () => {
                    return new Promise((resolve) => {
                        resolve('test');
                    });
                },
                render: () => {}
            };
        }
    });

    cy.get('#input-email').type('');
    cy.get('#input-password').type(''}');
    cy.get(':nth-child(6) > .k-button').should('be.visible').click({force : true})

});

2nd Solution:
/// <reference types = "cypress" />

const { HttpStatusCode } = require("axios");

// import 'cypress-iframe'
describe('reCaptcha Bypass', () => {
    beforeEach(() => {
        cy.intercept('POST', 'https://www.recaptcha.net/recaptcha/api2/', (req) => {
            console.log('Intercepted reCaptcha verification request', req);
            req.reply({
                HttpStatusCode: 200,
                body: {
                    success: true
                }
             });
        }).as('reCaptchaVerify');
    });

    it('Login test', function(){

        cy.visit('/')
        cy.get('#input-email').type('')
        cy.get('#input-password').type('{enter}')
        // cy.get('[style="margin: 0px auto; top: 0px; left: 0px; right: 0px; position: fixed; border: 1px solid rgb(204, 204, 204); z-index: 2000000000; background-color: rgb(255, 255, 255); width: 400px; height: 580px;"] > iframe').then($iframe => {
        //     const $body = $iframe.contents().find('body');
        //     cy.wrap($body).find('#recaptcha-token').click({force: true});
        // })
        cy.log('Waiting for reCaptcha verification');
        // cy.get('div[style*="position: fixed"]').invoke('remove');
        cy.get(':nth-child(6) > .k-button').should('be.visible').click({ force: true })
    
    });
});

Any help would be appreciated. Thank you!

**Comment by portyu9** [comment]
Hi @sejal-bansal 

I would not try to fake the real reCAPTCHA response with `cy.intercept()` or a mocked `window.grecaptcha`.

Google provides **official reCAPTCHA v2 test keys specifically for automated tests**:

```text
Site key:
6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI

Secret key:
6LeIxAcTAAAAAGG-vFI1TnRWxMZNFuojJ4WifJWe
```

Google documents that with these v2 test keys you will always receive **No CAPTCHA** and the verification request will pass. The widget displays a warning so the keys are not accidentally used in production.

Use the test site key in the frontend and the matching test secret in your backend for your Cypress/test environment.

Your current approaches have an important limitation:

```js
cy.intercept(...)
```

intercepts requests made through the browser under test. The important reCAPTCHA verification normally happens **server-to-server** from your application backend to Google's verification service, so intercepting browser requests cannot turn an invented token into a valid reCAPTCHA token.

Likewise, replacing:

```js
window.grecaptcha
```

can simulate your application's client-side callback behavior, but it does not create a token that Google's backend verification will accept.

So the preferred choices are:

1. use Google's official v2 test keys in the test environment; or
2. if you own the application architecture, abstract/mock the server-side CAPTCHA verification in your E2E test environment.

That lets Cypress test your application's behavior without trying to defeat a real CAPTCHA challenge.

No screenshots or attachments were included in this discussion.

Please mark this answer as accepted if it helped, thank you.
